woocommerce-services.php

```diff
@@ -249,6 +256,18 @@ class WC_Connect_Loader {
 
 		public static function plugin_deactivation() {
 			wp_clear_scheduled_hook( 'wc_connect_fetch_service_schemas' );
+
+			/**
+			 * When we deactivate the plugin after wcshipping_migration_state has started,
+			 * that means the migration is done. We can mark it as completed before we deactivate the plugin.
+			 */
+			$migration_state = WC_Connect_Options::get_option( 'wcshipping_migration_state' );
+			if ( ! class_exists(WC_Connect_API_Constants::class) ) {
+				require_once __DIR__ . '/classes/class-wc-connect-api-constants.php';
+			}
+			if ( $migration_state === WC_Connect_API_Constants::MIGRATION_STATE_STARTED ) {
```

Comment:
How about we have a different migration state after plugin installation? That way we're able to distinguish "started but not installed" (if for whatever reason it is not changed to "failed" inbetween) and "started and installed". Then we can transition to completed only from "installed".

Reply:
I drew out a finite state automata :D I think we can have 8 states and we can start from any of them and will always get to the end. This includes error handling as well. What do you think about adding this to woocommerce shipping issue 134? I can update that ticket to handle how we want to tackle each migration state in the automata?

![image](https://github.com/Automattic/woocommerce-services/assets/572862/4a552259-61d8-42b4-9687-159aa2f0e570)

Reply:
> we make sure we don't transition sites from "started" to "completed" when nothing really happened due to an error (and failure to update the state to "failed").

Yea 100%. This is also why I think we should tackle this problem in https://github.com/woocommerce/shipping/issues/134 instead. 134 deal with the whole error handling. I will update the description there so that the issue needs to detect which of these 8 (or more) states the migration was in, and carry on from there.